lib-ee/emqx_ee_connector/src/emqx_ee_connector_rocketmq.erl

```diff
@@ -0,0 +1,348 @@
+            }),
+            Result;
+        _ ->
+            ?tp(
+                rocketmq_connector_query_return,
+                #{result => Result}
+            ),
+            Result
+    end.
+
+safe_do_produce(InstanceId, QueryFunc, ClientId, TopicKey, Data, ProducerOpts, RequestTimeout) ->
+    try
+        Producers = get_producers(ClientId, TopicKey, ProducerOpts),
+        produce(InstanceId, QueryFunc, Producers, Data, RequestTimeout)
+    catch
+        _Type:Reason ->
+            {error, {unrecoverable_error, Reason}}
+    end.
+
+produce(_InstanceId, QueryFunc, Producers, Data, RequestTimeout) ->
+    rocketmq:QueryFunc(Producers, Data, RequestTimeout).
+
+parse_template(Config) ->
+    Templates =
+        case maps:get(template, Config, undefined) of
+            undefined -> #{};
+            <<>> -> #{};
+            Template -> #{send_message => Template}
+        end,
+
+    parse_template(maps:to_list(Templates), #{}).
+
+parse_template([{Key, H} | T], Templates) ->
+    ParamsTks = emqx_plugin_libs_rule:preproc_tmpl(H),
+    parse_template(
+        T,
+        Templates#{Key => ParamsTks}
+    );
+parse_template([], Templates) ->
+    Templates.
+
+get_topic_key({_, Msg}, RawTopic, TopicTks) ->
+    {RawTopic, emqx_plugin_libs_rule:proc_tmpl(TopicTks, Msg)};
+get_topic_key([Query | _], RawTopic, TopicTks) ->
+    get_topic_key(Query, RawTopic, TopicTks).
+
+apply_template({Key, Msg} = _Req, Templates) ->
+    case maps:get(Key, Templates, undefined) of
+        undefined ->
+            emqx_json:encode(Msg);
+        Template ->
+            emqx_plugin_libs_rule:proc_tmpl(Template, Msg)
+    end;
+apply_template([{Key, _} | _] = Reqs, Templates) ->
+    case maps:get(Key, Templates, undefined) of
+        undefined ->
+            [emqx_json:encode(Msg) || {_, Msg} <- Reqs];
+        Template ->
+            [emqx_plugin_libs_rule:proc_tmpl(Template, Msg) || {_, Msg} <- Reqs]
+    end.
+
+client_id(InstanceId) ->
+    erlang:list_to_atom("bridge_rocketmq:" ++ to_str(InstanceId)).
+
+to_str(List) when is_list(List) -> List;
+to_str(Bin) when is_binary(Bin) -> binary_to_list(Bin);
+to_str(Atom) when is_atom(Atom) -> atom_to_list(Atom).
+
+to_atom(List) when is_list(List) -> erlang:list_to_atom(List);
+to_atom(Bin) when is_binary(Bin) -> erlang:binary_to_atom(Bin);
+to_atom(Atom) when is_atom(Atom) -> Atom.
+
+redact(Msg) ->
+    emqx_misc:redact(Msg, fun is_sensitive_key/1).
+
+is_sensitive_key(security_token) ->
+    true;
+is_sensitive_key(_) ->
+    false.
+
+make_producer_opts(
+    InstanceId,
+    #{
+        username := Username,
+        password := Password,
+        security_token := SecurityToken,
+        send_buffer := SendBuff,
+        refresh_interval := RefreshInterval
+    }
+) ->
+    ACLInfo = acl_info(Username, Password, SecurityToken),
+    #{
+        tcp_opts => [{sndbuf, SendBuff}],
+        ref_topic_route_interval => RefreshInterval,
+        name => to_atom(InstanceId),
```

Comment:
Do we really need this to be an atom?

The instance ID will change each time a bridge is (re)created or even if the bridge is simply health probed, so this could lead to an atom leak.

Reply:
No,  According to here, https://github.com/emqx/rocketmq-client-erl/blob/e1975007ca55d60e1368853c15869edf6bf81beb/src/rocketmq_producers_sup.erl#L67, the `name` isn't necessary

Reply:
Sorry, I meant to check if we need `name` _to be an atom_, not if the parameter is required.

If it's possible to keep it as a binary, it would avoid atom leakage.

Reply:
Yes, I know. The name must be an atom, this is a limit for the driver, but the name not is required, so I delete it from the options map